File: lesson_12/structs_java/structs_app/src/main/java/com/codedifferently/lesson12/Stack.java
package com.codedifferently.lesson12;

/** Implement the below Stack by providing code for the class methods. */
public class Stack {
  private ListNode top;

  public Stack() {
    this.top = null;
  }

  public void push(int value) {
    ListNode topHat = new ListNode(value);
    topHat.next = top;
    top = topHat;
  }

  public int pop() {
    int headNumber = top.val;
    top = top.next;
    return headNumber;
  }

  public int peek() {
    return top.val;
  }

  public boolean isEmpty() {
    return top == null;
  }
}

File: lesson_12/structs_ts/src/stack.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { ListNode } from './list_node.js';

export class Stack {
  private top: ListNode | undefined;

  constructor() {
    this.top = undefined;
  }

  push(value: number): void {
    const newNode = new ListNode(value);
    newNode.next = this.top;
    this.top = newNode;
  }

  pop(): number | undefined {
    if (this.isEmpty()) return undefined;
    const value = this.top!.value;
    this.top = this.top!.next;
    return value;
  }

  peek(): number | null {
    return this.isEmpty() ? null : this.top?.value;
  }

  isEmpty(): boolean {
    return this.top === undefined;
  }
}
